pkg/manager/manager_test.go

```diff
@@ -0,0 +1,312 @@
+package manager
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/docker/infrakit/pkg/discovery"
+	"github.com/docker/infrakit/pkg/leader"
+	group_mock "github.com/docker/infrakit/pkg/mock/spi/group"
+	store_mock "github.com/docker/infrakit/pkg/mock/store"
+	group_rpc "github.com/docker/infrakit/pkg/rpc/group"
+	"github.com/docker/infrakit/pkg/rpc/server"
+	"github.com/docker/infrakit/pkg/spi/group"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+type testLeaderDetector struct {
+	t     *testing.T
+	me    string
+	input <-chan string
+	stop  chan struct{}
+	ch    chan leader.Leadership
+}
+
+func (l *testLeaderDetector) Start() (<-chan leader.Leadership, error) {
+	l.stop = make(chan struct{})
+	l.ch = make(chan leader.Leadership)
+	go func() {
+		for {
+			select {
+			case <-l.stop:
+				return
+			case found := <-l.input:
+				if found == l.me {
+					l.ch <- leader.Leadership{Status: leader.Leader}
+				} else {
+					l.ch <- leader.Leadership{Status: leader.NotLeader}
+				}
+			}
+		}
+	}()
+	return l.ch, nil
+}
+
+func (l *testLeaderDetector) Stop() {
+	close(l.stop)
+}
+
+func testEnsemble(t *testing.T,
+	dir, id string,
+	leader chan string,
+	ctrl *gomock.Controller,
+	configStore func(*store_mock.MockSnapshot),
+	configureGroup func(*group_mock.MockPlugin)) (Backend, server.Stoppable) {
```

Comment:
I think you're referring to the long list of parameters in the function `NewManager` in `manager.go`?

This function uses parameters that are of complex type (e.g. `discoverry.Plugin`, `leader.Detector`) instead of simple types like strings and bools (seen in the k8s you referenced), so while the function arg list is long, it's hard to mess things up without the compiler complaining.  

I could look into doing something with embedding / composition with a struct like a `manager.Config`; however, my concern is that the fields would be public/exported and can be mutated after a manager instance is initialized.  Having the dependencies passed in as function params allow the manager's dependencies (as struct fields) to be unexported and immutable after initialization.